lib/vssh/libssh2.c

```diff
@@ -2782,7 +2793,8 @@ static CURLcode ssh_statemach_act(struct connectdata *conn, bool *block)
 
     case SSH_SESSION_FREE:
 #ifdef HAVE_LIBSSH2_KNOWNHOST_API
-      if(sshc->kh) {
+      if(sshc->kh && !(data->share &&
+        data->share->specifier & (1<< CURL_LOCK_DATA_KNOWNHOSTS)) ) {
```

Comment:
Shouldn't the freeing of this how rather be done when the corresponding share object is freed? How do you prevent that this free isn't done while another handle us still using it?